uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/implementation/addon/EventStorageImpl.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.testing.mqtt.client.control.implementation.addon;
+
+import com.aws.greengrass.testing.mqtt.client.control.api.addon.Event;
+import com.aws.greengrass.testing.mqtt.client.control.api.addon.EventFilter;
+import com.aws.greengrass.testing.mqtt.client.control.api.addon.EventStorage;
+import lombok.NonNull;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation storage of the events.
+ */
+public class EventStorageImpl implements EventStorage {
+
+    private final List<Event> events;
```

Comment:
Why not store the events on the Agent (or Connection) itself? Then we would not need to filter the events by which agent it came from, only need to lookup the Agent and get the list of events from it.

Reply:
The goal of the project does not include testing MQTT broker compliance, testing shared subscriptions is not needed. Even if it is needed, it is easy to validate using the procedural implementation that I described earlier (not using your object oriented event filter approach).

Reply:
We obligated to do that test in Phase 2 of the project. Please check actual SoW document here https://docs.google.com/document/d/1ILcnkuCpK_SoUclyOF4GAjStoyJOUqGD  5.4  point 5. x.  about shared subscriptions

Reply:
You do not need to validate the broker implements shared subscriptions properly, only that you are able to publish to a shared subscription topic and receive the message in a subscriber to the shared subscription topic.